go.mod

```diff
@@ -422,6 +422,7 @@ replace (
 	github.com/prometheus/client_model => github.com/prometheus/client_model v0.0.0-20190129233127-fd36f4220a90
 	github.com/prometheus/common => github.com/prometheus/common v0.4.1
 	github.com/prometheus/procfs => github.com/prometheus/procfs v0.0.2
+	github.com/prometheus/prometheus => github.com/prometheus/prometheus v2.5.0+incompatible
```

Comment:
this is a very large dependency to add for a single linter file. can we replicate the checks it is doing ourselves and avoid this?

Reply:
Good point and technically yes.  The other two dependencies are in go.mod already.
```
dto "github.com/prometheus/client_model/go"
"github.com/prometheus/common/expfmt"
```

@brancz How do you think?

Reply:
doesn't prometheus/prometheus import many other dependencies? That is somewhat masked because it does not currently declare a go.mod file, but once it does, we would pick up all of those as explicit transitive dependencies. For such a small benefit, it would seem to be better just to replicate the 2-3 checks

Reply:
Maybe instead of using promtool as a library we can try and find a way to use the binary for our checks? I would prefer if we didn't have to sync these things manually, that's kind of the point of the tool's existence. I tend to agree that the dependency might be a bit exaggerated. Definitions of `go.mod` files for Prometheus to be consumed as mods is still being figured out, I don't necessarily see this progressing very quickly for what it's worth.

Reply:
@mdlayher proposed to move `promlint` to `github.com/prometheus/common`.  But @brian-brazil stopped the progress. I think we should reconsider [this proposal](https://github.com/prometheus/prometheus/issues/5317).

Reply:
> I tend to agree that the dependency might be a bit exaggerated.

I disagree. This is what prometheus/prometheus would bring in:

direct dependency on things we already depend on (would complicate version selection):
github.com/Azure/azure-sdk-for-go
github.com/Azure/go-autorest
github.com/aws/aws-sdk-go
github.com/alecthomas/units
github.com/go-kit/kit
github.com/go-logfmt/logfmt
github.com/go-openapi/strfmt
github.com/gogo/protobuf
github.com/google/pprof
github.com/gophercloud/gophercloud
github.com/grpc-ecosystem/grpc-gateway
github.com/json-iterator/go
github.com/miekg/dns
github.com/mwitkow/go-conntrack
github.com/pkg/errors
github.com/prometheus/client_golang
github.com/prometheus/client_model
github.com/prometheus/common
github.com/soheilhy/cmux
golang.org/x/net/netutil
golang.org/x/oauth2/google
golang.org/x/sync/errgroup
golang.org/x/sys/unix
golang.org/x/time/rate
google.golang.org/api/compute/v1
google.golang.org/api/option
google.golang.org/genproto/googleapis/api/annotations
google.golang.org/grpc
gopkg.in/alecthomas/kingpin.v2
gopkg.in/yaml.v2
k8s.io/klog

direct dependency on things we don't currently depend on (would pull in all their transitive dependencies for version computation):
github.com/cespare/xxhash
github.com/cespare/xxhash/v2
github.com/edsrzf/mmap-go
github.com/golang/snappy
github.com/hashicorp/consul
github.com/influxdata/influxdb
github.com/oklog/run
github.com/oklog/ulid
github.com/opentracing-contrib/go-stdlib/nethttp
github.com/opentracing/opentracing-go
github.com/prometheus/alertmanager
github.com/samuel/go-zookeeper
github.com/shurcooL/httpfs
github.com/shurcooL/vfsgen
gopkg.in/fsnotify/fsnotify.v1

direct dependency on things published out of the staging repo (adds a circular dependency on kubernetes packages that can gridlock staging changes or dependency updates):
k8s.io/api
k8s.io/apimachinery
k8s.io/client-go

The last category (dependencies on things published from staging) is the most significant issue, but adding things like github.com/influxdata/influxdb, github.com/samuel/go-zookeeper and github.com/hashicorp/consul into the dependency tree is also a heavy price to pay for such a small utility package.

Reply:
> Maybe instead of using promtool as a library we can try and find a way to use the binary for our checks?

@brancz 
I've no idea about how to use a binary of promtool.

By the way, how do you think about moving `promtool` to `client_golang/prometheus/testutil`? 
The previous discussion is [here](https://github.com/prometheus/prometheus/issues/5317#issuecomment-571129338).

Reply:
testutil is for our internal testing libraries, if has nothing to do with the promtool CLI.

Reply:
How about `client_golang/prometheus/promlint`?  @brian-brazil

@brancz I'd like to set up this check in v1.18. We need to hurry up.

Thanks.

Reply:
You'd have to ask @beorn7 if he'd be willing to expose promlint as a public API as part of client_golang.

I would however suggest that using promtool as a binary is the better option.

Reply:
> I would however suggest that using promtool as a binary is the better option.

Yes, I agree. But I have no clue about how to do that. @brancz can you explain more?
This is the fastest way to push this forward, we don't need to wait for a new release if there is a public API.

Reply:
Given how Go Modules work, I'm hesitant to put all kind of "client" and "public API" functionality in client_golang. It is really the instrumentation client, and due to an accident, the repo also contains a half-abandoned HTTP client, which is already causing pain.